src/vmm/src/gdb/arch/aarch64.rs

```diff
@@ -1,62 +1,297 @@
-    unimplemented!()
+    let mut dbg = kvm_guest_debug {
+        control,
+        ..Default::default()
+    };
+
+    for (i, addr) in addrs.iter().enumerate() {
+        // DBGBCR_EL1 (Debug Breakpoint Control Registers, D13.3.2):
+        // bit 0: 1 (Enabled)
+        // bit 1~2: 0b11 (PMC = EL1/EL0)
+        // bit 5~8: 0b1111 (BAS = AArch64)
+        // others: 0
+        dbg.arch.dbg_bcr[i] = 0b1 | (0b11 << 1) | (0b1111 << 5);
+        // DBGBVR_EL1 (Debug Breakpoint Value Registers, D13.3.3):
+        // bit 2~52: VA[2:52]
+        dbg.arch.dbg_bvr[i] = (!0u64 >> 11) & addr.0;
+    }
+
+    vcpu_fd.set_guest_debug(&dbg)?;
+
+    Ok(())
+}
+
+/// Bits in a Vcpu pstate for IRQ
+const IRQ_ENABLE_FLAGS: u64 = 0x80 | 0x40;
+/// Register id for pstate
+const PSTATE_ID: u64 = arm64_core_reg_id!(KVM_REG_SIZE_U64, offset_of!(user_pt_regs, pstate));
+
+/// Disable IRQ interrupts to avoid getting stuck in a loop while single stepping
+///
+/// When GDB hits a single breakpoint and resumes it will follow the steps:
+///  - Clear SW breakpoint we've hit
+///  - Single step
+///  - Re-insert the SW breakpoint
+///  - Resume
+/// However, with IRQ enabled the single step takes us into the IRQ handler so when we resume we
+/// immediately hit the SW breapoint we just re-inserted getting stuck in a loop.
+fn toggle_interrupts(vcpu_fd: &VcpuFd, enable: bool) -> Result<(), GdbTargetError> {
+    let mut pstate = get_sys_reg(PSTATE_ID, vcpu_fd)?;
+
+    if enable {
+        pstate |= IRQ_ENABLE_FLAGS;
+    } else {
+        pstate &= !IRQ_ENABLE_FLAGS;
+    }
+
+    vcpu_fd.set_one_reg(pstate_id, &pstate.to_le_bytes())?;
+
+    Ok(())
 }
 
 /// Configures the Vcpu for debugging and sets the hardware breakpoints on the Vcpu
 pub fn vcpu_set_debug(
-    _vcpu_fd: &VcpuFd,
-    _addrs: &[GuestAddress],
-    _step: bool,
+    vcpu_fd: &VcpuFd,
+    addrs: &[GuestAddress],
+    step: bool,
 ) -> Result<(), GdbTargetError> {
-    unimplemented!()
+    let mut control = KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_USE_HW | KVM_GUESTDBG_USE_SW_BP;
+    if step {
+        control |= KVM_GUESTDBG_SINGLESTEP;
+    }
+
+    toggle_interrupts(vcpu_fd, step)?;
+    set_kvm_debug(control, vcpu_fd, addrs)
 }
 
-/// Injects a BP back into the guest kernel for it to handle, this is particularly useful for the
-/// kernels selftesting which can happen during boot.
+/// KVM does not support injecting breakpoints on aarch64 so this is a no-op
 pub fn vcpu_inject_bp(
     _vcpu_fd: &VcpuFd,
     _addrs: &[GuestAddress],
     _step: bool,
 ) -> Result<(), GdbTargetError> {
-    unimplemented!()
+    Ok(())
 }
 
+/// The number of core registers we read from the Vcpu
+const CORE_REG_COUNT: usize = 33;
+/// Stores the register ids of registers to be read from the Vcpu
+const CORE_REG_IDS: [u64; CORE_REG_COUNT] = {
+    let mut regs = [0; CORE_REG_COUNT];
+    let mut idx = 0;
+
+    let reg_offset = offset_of!(kvm_regs, regs);
+    let mut off = reg_offset;
+    while idx < 31 {
```

Comment:
can have a constant defined for 31 (I can see it used in multiple places).

Reply:
Good call, I've added this as a constant now.